server/src/main/java/org/opensearch/rest/action/cat/RestShardsAction.java

```diff
@@ -453,7 +498,106 @@ Table buildTable(RestRequest request, ClusterStateResponse state, IndicesStatsRe
 
             table.endRow();
         }
-
         return table;
     }
+
+    private List<String> getListOfIndicesSortedByCreateTime(final ClusterStateResponse clusterStateResponse) {
+        List<String> indicesList = new ArrayList<String>(clusterStateResponse.getState().getRoutingTable().getIndicesRouting().keySet());
+        indicesList.sort((index1, index2) -> {
+            Long index1CreationTimeStamp = clusterStateResponse.getState().metadata().indices().get(index1).getCreationDate();
+            Long index2CreationTimeStamp = clusterStateResponse.getState().metadata().indices().get(index2).getCreationDate();
+            if (index1CreationTimeStamp.equals(index2CreationTimeStamp)) {
+                return index1.compareTo(index2);
+            }
+            return (index1CreationTimeStamp - index2CreationTimeStamp) > 0 ? 1 : -1;
+        });
+        return indicesList;
+    }
+
+    private String getNextTokenForPaginatedResponse(
+        final RestRequest request,
+        ClusterStateResponse clusterStateResponse,
+        List<String> indicesToBeQueried,
+        List<ShardRouting> shardRoutingResponseList
+    ) {
+        final long defaultPageSize = (long) clusterStateResponse.getState().nodes().getDataNodes().size() + 1;
+
+        // Get the nextToken provided in the request
+        final String nextTokenInRequest = Objects.equals(request.param("nextToken"), "null")
+            ? null
+            : new String(Base64.getDecoder().decode(request.param("nextToken")), UTF_8);
+
+        List<String> sortedIndicesList = getListOfIndicesSortedByCreateTime(clusterStateResponse);
+
+        // Since all the shards for last ID would have already been sent in the last response,
+        // start iterating from the next shard for current page
+        int newPageStartShardID = nextTokenInRequest == null ? 0 : Integer.parseInt(nextTokenInRequest.split("\\$")[0]) + 1;
+
+        // Since all the shards corresponding to the last processed index might not have been included in the last page,
+        // start iterating from the last index number itself
+        int newPageStartIndexNumber = nextTokenInRequest == null ? 0 : Integer.parseInt(nextTokenInRequest.split("\\$")[1]);
+
+        // Get the number of shards upto the maxPageSize
+        long shardCountSoFar = 0L;
+        int lastProcessedShardNumber = -1;
+        int lastProcessedIndexNumber = -1;
```

Comment:
Looking forward to new paginated tests covering all use cases. You will need to add UTs and REST tests. Also, you will need to make changes in https://github.com/opensearch-project/opensearch-api-specification